qiskit_nature/utils/linalg.py

```diff
@@ -179,3 +179,136 @@ def _swap_columns(matrix: np.ndarray, i: int, j: int) -> None:
     column_i = matrix[:, i].copy()
     column_j = matrix[:, j].copy()
     matrix[:, i], matrix[:, j] = column_j, column_i
+
+
+def modified_cholesky(
+    mat: np.ndarray, *, error_threshold: float = 1e-8, max_vecs: int | None = None
+) -> np.ndarray:
+    r"""Modified Cholesky decomposition.
+
+    The modified Cholesky decomposition of a square matrix :math:`M` has the form
+
+    .. math::
+        M = \sum_{i=1}^N v_i v_i^\dagger
+
+    where each :math:`v_i` is a vector. :math:`M` must be positive definite.
+    No checking is performed to verify whether :math:`M` is positive definite.
+    The number of terms :math:`N` in the decomposition depends on the allowed
+    error threshold. A larger error threshold may yield a smaller number of terms.
+    Furthermore, the ``max_vecs`` parameter specifies an optional upper bound
+    on :math:`N`. The ``max_vecs`` parameter is always respected, so if it is
+    too small, then the error of the decomposition may exceed the specified
+    error threshold.
+
+    .. note::
+        No checking is performed to verify whether the input matrix is positive definite.
+        If the input matrix is not positive definite, then the decomposition returned will be invalid.
+
+    References:
+        - `arXiv:1711.02242`_
+        - `International Journal of Quantum Chemistry 12, 683 (1977)`_
+
+    Args:
+        mat: The matrix to decompose.
+        error_threshold: Threshold for allowed error in the decomposition.
+            The error is defined as the maximum absolute difference between
+            an element of the original tensor and the corresponding element of
+            the reconstructed tensor.
+        max_vecs: The maximum number of vectors to include in the decomposition.
+
+    Returns:
+        The Cholesky vectors ``v_i`` assembled into a 2-dimensional Numpy array
+        whose columns are the vectors.
+
+    .. _arXiv:1711.02242: https://arxiv.org/abs/1711.02242
+    .. _International Journal of Quantum Chemistry 12, 683 (1977): https://doi.org/10.1002/qua.560120408
+    """
+    dim, _ = mat.shape
+
+    if max_vecs is None:
+        max_vecs = dim
+
+    cholesky_vecs = np.zeros((dim, max_vecs + 1), dtype=mat.dtype)
+    errors = np.real(np.diagonal(mat).copy())
+    for index in range(max_vecs + 1):
+        max_error_index = np.argmax(errors)
+        max_error = errors[max_error_index]
+        if max_error < error_threshold:
+            break
+        cholesky_vecs[:, index] = mat[:, max_error_index]
+        if index:
+            cholesky_vecs[:, index] -= (
+                cholesky_vecs[:, 0:index] @ cholesky_vecs[max_error_index, 0:index].conj()
+            )
+        cholesky_vecs[:, index] /= np.sqrt(max_error)
+        errors -= np.abs(cholesky_vecs[:, index]) ** 2
+
+    return cholesky_vecs[:, :index]
+
+
+def double_factorized(
+    two_body_tensor: np.ndarray, *, error_threshold: float = 1e-8, max_vecs: int | None = None
```

Comment:
Same here, thinking about `sparse.SparseArray`. The output of this method can get _very_ large, so supporting some sparse tensor format is desirable.

Reply:
It would fail with the same message as above.